rskj-core/src/main/java/co/rsk/net/handler/quota/TxQuotaChecker.java

```diff
@@ -0,0 +1,189 @@
+        updateLastBlockGasLimit(currentContext.bestBlock.getGasLimitAsInteger());
+
+        TxQuota senderQuota = updateQuota(newTx, true, currentContext);
+
+        boolean existsDestination = currentContext.repository.getAccountState(newTx.getReceiveAddress()) != null;
+        boolean isEOA = !currentContext.repository.isContract(newTx.getReceiveAddress());
+        if (!existsDestination || isEOA) {
+            updateQuota(newTx, false, currentContext);
+        }
+
+        double consumedVirtualGas = calculateConsumedVirtualGas(newTx, replacedTx, currentContext);
+
+        return senderQuota.acceptVirtualGasConsumption(consumedVirtualGas, newTx);
+    }
+
+    /**
+     * Cleans from the underlying map those entries that have <code>maxQuota</code> after being refreshed
+     * This method is intended to be called periodically with a rate similar to the time needed for an account to get <code>maxQuota</code>
+     */
+    public synchronized void cleanMaxQuotas() {
+        if (lastBlockGasLimit == -1) {
+            // no transactions yet processed
+            return;
+        }
+
+        long maxGasPerSecond = getMaxGasPerSecond(lastBlockGasLimit);
+        long maxQuota = getMaxQuota(maxGasPerSecond);
+
+        logger.debug("Clearing quota map, size before {}", this.accountQuotas.size());
+
+        Iterator<Map.Entry<RskAddress, TxQuota>> quotaIterator = accountQuotas.entrySet().iterator();
+        while (quotaIterator.hasNext()) {
+            Map.Entry<RskAddress, TxQuota> quotaEntry = quotaIterator.next();
+            RskAddress address = quotaEntry.getKey();
+            TxQuota quota = quotaEntry.getValue();
+            double accumulatedVirtualGas = quota.refresh(address, maxGasPerSecond, maxQuota);
+            boolean maxQuotaGranted = BigDecimal.valueOf(maxQuota).compareTo(BigDecimal.valueOf(accumulatedVirtualGas)) == 0;
+            if (maxQuotaGranted) {
+                logger.trace("Clearing {}, it has maxQuota", quota);
+                quotaIterator.remove();
+            }
+        }
+
+        logger.debug("Clearing quota map, size after {}", this.accountQuotas.size());
+    }
+
+    public TxQuota getTxQuota(RskAddress address) {
+        return this.accountQuotas.get(address);
+    }
+
+    private TxQuota updateQuota(Transaction newTx, boolean isTxSource, CurrentContext currentContext) {
+        BigInteger blockGasLimit = currentContext.bestBlock.getGasLimitAsInteger();
+        long maxGasPerSecond = getMaxGasPerSecond(blockGasLimit.longValue());
+        long maxQuota = getMaxQuota(maxGasPerSecond);
+
+        RskAddress address = isTxSource ? newTx.getSender() : newTx.getReceiveAddress();
+
+        TxQuota quotaForAddress = this.accountQuotas.get(address);
+        if (quotaForAddress == null) {
+            long accountNonce = currentContext.state.getNonce(address).longValue();
```

Comment:
I had that same doubt, I didn't like too much "switching" to `long`.

But, since the size of a `long` is enough by far for our needs (correct me if I'm wrong), finally I've decided to use it for the sake of code readability: `TxVirtualGasCalculator` has tons of operations and IMO its logic is easier to understand using `long`, it was a bit chaotic when I first tried using `BigInteger`, `BigDecimal`, etc.

I thought that being this change encapsulated just in `TxVirtualGasCalculator`, it was a not so problematic switch while we benefit from code readability.

Reply:
my concern is not only about `nonce`. In fact, as I understand, there should not be a lot of math around `nonce`'s

you're probably referring to [this](https://web.archive.org/web/20120601021936/http://firstclassthoughts.co.uk/java/traps/big_decimal_traps.html) thing ) 

my concern is regarding [this](https://web.archive.org/web/20120529055141/http://firstclassthoughts.co.uk/java/traps/java_double_traps.html). 

There's a lot of "math" with gas units, sizes etc., for instance, [here](https://github.com/rsksmart/rskj/blob/4c8c8cb39abede4d32136c7d35aa50b8bfda2719/rskj-core/src/main/java/co/rsk/net/handler/quota/TxVirtualGasCalculator.java#L31-L47), so approximations of `double`'s may accumulate on the way.

Reply:
Yep, I was aware of some of those pitfalls, but being ~15 the precision of a double and the usage we are doing of them (no equal comparison, etc.), do you think it is an actual problem that accumulation or anything else?

I am just trying to improve readability, which is clearly decreased when using BigDecimal and BigInteger, but I could be failing to spot a potential problem...

Reply:
As agreed with Vlad, unless security has any concern, we will move forward with the current implementation.